SPECS/tdnf/virtual-repo-snapshot.patch

```diff
@@ -0,0 +1,906 @@
++// filter general error
++# define ERROR_TDNF_TIME_FILTER_GENERAL      (ERROR_TDNF_TIME_FILTER_BASE + 4)
+ 
+ #define ERROR_TDNF_PLUGIN_BASE          2000
+ 
+diff --git a/include/tdnftypes.h b/include/tdnftypes.h
+index a806010..e638427 100644
+--- a/include/tdnftypes.h
++++ b/include/tdnftypes.h
+@@ -9,6 +9,7 @@
+ #pragma once
+ 
+ #include <curl/curl.h>
++#include <time.h>
+ 
+ #ifdef __cplusplus
+ extern "C" {
+@@ -257,6 +258,7 @@ typedef struct _TDNF_CONF
+     int nCheckUpdateCompat;
+     int nDistroSyncReinstallChanged;
+     char* pszRepoDir;
++    char* pszRepoEpochTimeLimit;
+     char* pszCacheDir;
+     char* pszPersistDir;
+     char* pszProxy;
+diff --git a/solv/defines.h b/solv/defines.h
+index 38f5ab1..ebc7bd0 100644
+--- a/solv/defines.h
++++ b/solv/defines.h
+@@ -18,4 +18,32 @@
+         }                                                          \
+     } while(0)
+ 
++typedef struct {
++    // frequently changed values
++    char * pszElementBuffer;
++    int nBufferLen;
++    int nInPackage;
++    int nPrintPackage;
++    int nTimeFound;
++
++    // managed values
++    int nBufferMaxLen;
++    int nDepth;
++    int nPrevElement; // enum 0 -> start, 1 -> data, 2 -> end
++
++    //set and forget on creation
++    time_t nSearchTime;
++    FILE * pbOutfile;
++} TDNFFilterData;
++
++#define TDNF_MAX_FILTER_INPUT_THRESHOLD 10000000
++
++#define BAIL_ON_TDNF_TIME_FILTER_ERROR(dwError) \
++    do {                                                           \
++        if (dwError)                                               \
++        {                                                          \
++            goto error;                                            \
++        }                                                          \
++    } while(0)
++
+ #endif /* __SOLV_DEFINES_H__ */
+diff --git a/solv/includes.h b/solv/includes.h
+index 2ab0c5c..1e39ecb 100644
+--- a/solv/includes.h
++++ b/solv/includes.h
+@@ -10,7 +10,6 @@
+ #include <unistd.h>
+ #include <stdbool.h>
+ #include <sys/types.h>
+-#include <sys/stat.h>
+ #include <dirent.h>
+ 
+ // libsolv
+@@ -44,4 +43,9 @@
+ #include "../history/history.h"
+ #include "prototypes.h"
+ 
++#include <expat.h>
++#include <stdlib.h>
++#include <time.h>
++#include <string.h>
++
+ #endif /* __SOLV_INCLUDES_H__ */
+diff --git a/solv/prototypes.h b/solv/prototypes.h
+index 2633b5e..0f5df8d 100644
+--- a/solv/prototypes.h
++++ b/solv/prototypes.h
+@@ -517,7 +517,8 @@ SolvReadYumRepo(
+     const char *pszPrimary,
+     const char *pszFilelists,
+     const char *pszUpdateinfo,
+-    const char *pszOther
++    const char *pszOther,
++    char * pszRepoEpochLimit
+     );
+ 
+ uint32_t
+diff --git a/solv/tdnfrepo.c b/solv/tdnfrepo.c
+index c27b907..aef6c61 100644
+--- a/solv/tdnfrepo.c
++++ b/solv/tdnfrepo.c
+@@ -9,6 +9,522 @@
+ 
+ #include "includes.h"
+ 
++// #### XML FILTER CODE ####
++void checkAndResizeBuffer(char ** pszCharBuffer, int * nBufferMaxLen, int nLengthToAdd) {
```

Comment:
It looks like nBufferMaxLen is not updated, but is passed as a pointer.

So...either null check the pointer...or pass nBufferMaxLen by value.

Also probably want to check that it's positive.

Reply:
inside of the loop, prior to the realloc the value is update. But yes I should check the value is positive :)